internals/daemon/daemon.go

```diff
@@ -203,18 +213,20 @@ func (c *Command) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 	if rsp, ok := rsp.(*resp); ok {
 		st := c.d.state
-		st.Lock()
-		_, rst := restart.Pending(st)
-		st.Unlock()
-		switch rst {
-		case restart.RestartSystem:
-			rsp.transmitMaintenance(errorKindSystemRestart, "system is restarting")
-		case restart.RestartDaemon:
-			rsp.transmitMaintenance(errorKindDaemonRestart, "daemon is restarting")
-		case restart.RestartSocket:
-			rsp.transmitMaintenance(errorKindDaemonRestart, "daemon is stopping to wait for socket activation")
+		if includeMaintenance {
+			st.Lock()
+			_, rst := restart.Pending(st)
+			st.Unlock()
```

Comment:
Since `restart.Pending()` already returns a `bool` as first return value (that's ignored here), what about tracking an `atomic.Bool` on `State`, so that checking whether `restart.Pending()` would return any of `RestartSystem`, `RestartDaemon` or `RestartSocket` would be lockless via the atomic bool (it will be false in > 99%-ish of cases?), and only when the atomic bool is actually true you take out the lock and request the reason with `restart.Pending(st)` with the lock held.

Having a quick look, it's the `restartStateKey{}` in `restart.go` that accesses the state, so tracking the access there should make it possible as a side-effect to track a boolean in `st`.

Instead of `if includeMaintenance {` above, this would become something like `if st.atomicBoolRestartPending {`.

Alternatively, because of `type RestartType int` you could also decide to store the full restart type as an `atomic.Int32` (32-bit ought to be enough to enumerate all possible restart types) and make this part here completely lock-free.

That would at least avoid adding `includeMaintenance` for the restart part.